### ISSUE TITLE
Chunked PLP transfer

### ISSUE THREAD

**Issue by agalakhov** [steffengy/tiberius#61]
Hi,

I'm working on PLP support for Varchar and now I'm refactoring the NVarchar support. Looks like there is no handling for `NotReady` in NVarchar reading. That is, if the data is not yet in the buffer, we'll return an error instead of NotReady, which is incorrect. Is this true or am I missing something?

Regards,
Alexey

**Comment by steffengy** [comment]
Great that you're taking that up, the code in that part is really ... improvable.
With async/await that would be so much nicer :(

Packets are read into an underlying buffer,  
which always has a size of `n (amount of packets) * packetSize + r`.
`r` is the amount of remaining bytes (size of the last packet, which is the only one that is allowed to be differently sized than the negotiated packet size).

So I used a bunch of read calls that mostly work, and if not (e.g. a token spanning a boundary of the a single packet), those are ["translated"](https://github.com/steffengy/tiberius/blob/a2d81efff3c78b85722b8608d764e3040413be9b/tiberius/src/transport.rs#L476) since we know if the buffer is empty, we need more packets.

**So in summary, the `NotReady` case is implicitly handled by any `read_*` call
which might fail and indicate that more data is needed.**

**Comment by steffengy** [comment]
References #60

**Comment by steffengy** [comment]
The reason for that is historical:
- I had a more synchronous version with a lot of synchronous read calls (which are more readable),
and introduced this technique to get that to work as expected.
- Then I added some smarter asynchronous implementations that e.g. check if a string of length 200 is read, if the underlying buffer contains that much data (before trying to read it).
There I found returning `Async::NotReady` to be cleaner. It also allows to distinguish different sources of failure.

> I feel really bad about returning Err from a function that returns Poll

Poll is just `Result<Async<T>, E>` which currently would map to:
- `Ok(Async::Ready(T))`: all good, we got a result
- `Ok(Async::NotReady)`: we definitely need more data
- `Err(UnexpectedEof)`: we need more data (I do not remember a difference, probably just the historic reason above)
- `Err(err)`: something went wrong

**So I'd make `PLPReader` return Poll as well (atleast it's consistent then).**

#### Future
If I'd do it again, I'd probably just return Result<>, especially since with async/await,
that problem of explicitly creating those suspension points disappears.

Let me know what you think!

**Comment by agalakhov** [comment]
Thank you. The PLPReader will return `Poll`, and I'm thinking about using `Poll` for i.e. `u32` readers as well (and using `AsyncRead` everywhere, respectively).
